tobias/utils/motifs.py

```diff
@@ -248,70 +278,52 @@ def as_string(self, output_format="pfm"):
 			Motif format ("pfm", "jaspar", "meme")
 		"""
 
-		bases = ["A", "C", "G", "T"]
 		out_string = ""
 
-		#Establish which output format
-		if output_format in ["pfm", "jaspar"]:
-			for motif in self:
-				out_string += ">{0}\t{1}\n".format(motif.id, motif.name)
-
-				#Convert counts to string
-				counts_type = float if float in [type(element) for element in motif.counts[0]] else int	#Estimate from first base counts
-				if counts_type == float:
-					counts_string = [["{0:.5f}".format(element) for element in row_counts] for row_counts in motif.counts]
-				else:
-					counts_string = [[str(element) for element in row_counts] for row_counts in motif.counts]
-					element_width = max(sum([[len(element) for element in row_counts] for row_counts in counts_string], []))	#Length of longest element to align to
-					counts_string = [[" " * (element_width - len(element)) + element for element in row_counts] for row_counts in counts_string]
-
-				#Write out to format
-				for i, base_counts in enumerate(counts_string):
-					out_string += "{0} [ {1} ]\n".format(bases[i], "  ".join(base_counts)) if output_format == "jaspar" else "  ".join(base_counts) + "\n"
-				out_string += "\n"
+		header = True
+		for motif in self:
+			out_string += motif.as_string(output_format=output_format, header=header)
+			header = False
 
-		elif output_format == "meme":
-			
-			meme_header = "MEME version 4\n\n"
-			meme_header += "ALPHABET= ACGT\n\n"
-			meme_header += "strands: + -\n\n"
-			meme_header += "Background letter frequencies\n"
-			meme_header += " ".join(["{0} {1}".format(bases[i], self.bg[i]) for i in range(4)]) + "\n\n"
-			out_string += meme_header
-
-			for motif in self:
-				out_string += "MOTIF {0} {1}\n".format(motif.id, motif.name)
-				out_string += "letter-probability matrix: "
-				out_string += " ".join("{0}= {1}".format(key, value) for key, value in motif.info.items()) + "\n"
-
-				for i in range(motif.length):
-					row = [float(motif.counts[j][i]) for j in range(4)] 	#row contains original row from content
-					n_sites = round(sum(row), 0)
-					row_freq = ["{0:.6f}".format(num/n_sites) for num in row] 
-					out_string += " " + "  ".join(row_freq) + "\n"
-				
-				out_string += "\n"
-
-		#elif output_format == "transfac":
-		#	for motif in self:
-		#		out_string += motif.to_transfac()
+		return(out_string)
 
-		else:
-			raise ValueError("Format " + output_format + " is not supported")	
+	def get_background(self):
+		"""
+		Combines background of all motifs to a global background.
+  
+		Returns:
+			numpy array of frequencies
+		"""
+		total_n = 0
+		global_bg = None
+  
+		for motif in self:
+			if global_bg is None:
+				global_bg = np.array([0] * len(motif.bg), dtype=float)

```

Comment:
I don't understand this... you set the global_bg to None, then you ask whether it is None, and then you set it again? Maybe just initialize it at line 298? :-P

Reply:
But the length of self[0].bg is always 4.. or not? Should we support other alphabets than ATCG? But yea, an empty motif list could be a problem. Maybe a check for len(self) would solve it?

Reply:
At the moment the length is always 4 but I wanted to write a function that works regardless of length. Maybe that avoids some headaches in the future. In general, I tried to write so that other alphabets can be integrated as easily as possible.

Anyway, I will move the if above the loop and check for the length that should be more readable.

Reply:
Hmm, yes you are right to write it more general 👍 Good solution, I think my initial problem was also just confusion, so it is not a major problem as long as it is clear what happens.